controllers/storagecluster/odfinfoconfig.go

```diff
@@ -0,0 +1,293 @@
+	mutex.Lock()
+	defer mutex.Unlock()
+	if len(odfInfoConfigMap.Data) > 1 {
+		odfInfoKeyName := obj.getOdfInfoKeyName(storageCluster)
+		delete(odfInfoConfigMap.Data, odfInfoKeyName)
+		if err = r.Client.Update(r.ctx, odfInfoConfigMap); err != nil {
+			r.Log.Error(err, "Failed to update odf-info config map with deleted key.", "ConfigMap", klog.KRef(odfInfoConfigMap.Namespace, odfInfoConfigMap.Name), "Key", odfInfoKeyName)
+			return reconcile.Result{}, fmt.Errorf("failed to delete key %v in odf-info %v: %v", odfInfoKeyName, odfInfoConfigMap.Name, err)
+		}
+	} else {
+		if err = r.Client.Delete(r.ctx, odfInfoConfigMap); err != nil && !errors.IsNotFound(err) {
+			r.Log.Error(err, "Failed to delete odf-info config map.", "ConfigMap", klog.KRef(odfInfoConfigMap.Namespace, odfInfoConfigMap.Name))
+			return reconcile.Result{}, fmt.Errorf("failed to delete odf-info %v: %v", odfInfoConfigMap.Name, err)
+		}
+	}
+	return reconcile.Result{}, nil
+}
+
+func getOdfInfoData(r *StorageClusterReconciler, storageCluster *ocsv1.StorageCluster, clients string) (string, error) {
+	odfVersion, err := getOdfVersion(r, storageCluster)
+	if err != nil {
+		return "", err
+	}
+	cephFSId, err := getCephFsid(r, storageCluster)
+	if err != nil {
+		return "", err
+	}
+
+	odfDeploymentType := odfDeploymentTypeExternal
+	isDROptimized := false
+	// Set isDROptmized to "false" in case of external clusters as we currently don't have to way to determine
+	// if external cluster OSDs are using bluestore-rdr
+	if !storageCluster.Spec.ExternalStorage.Enable {
+		odfDeploymentType = odfDeploymentTypeInternal
+		if isDROptimized, err = getIsDROptimized(r, storageCluster); err != nil {
+			return "", err
+
+		}
+	}
+	var storageSystemName string
+	if storageSystemName, err = getStorageSystemName(storageCluster); err != nil {
+		return "", err
+	}
+
+	connectedClients, err := getConnectedClients(r, storageCluster, clients)
+	if err != nil {
+		return "", err
+	}
+
+	data := OdfInfoData{
+		Version:           odfVersion,
+		DeploymentType:    odfDeploymentType,
+		IsDROptimized:     isDROptimized,
+		StorageSystemName: storageSystemName,
+		Clients:           connectedClients,
+		StorageCluster: infoStorageCluster{
+			NamespacedName:          client.ObjectKeyFromObject(storageCluster),
+			StorageProviderEndpoint: storageCluster.Status.StorageProviderEndpoint,
+			CephClusterFSID:         cephFSId,
+		},
+	}
+	yamlData, err := yaml.Marshal(data)
+	if err != nil {
+		return "", err
+	}
+	return string(yamlData), nil
+
+}
+
+func getConnectedClients(r *StorageClusterReconciler, storageCluster *ocsv1.StorageCluster, clients string) ([]connectedClient, error) {
+	storageConsumers := &v1alpha1.StorageConsumerList{}
+	err := r.Client.List(r.ctx, storageConsumers, client.InNamespace(storageCluster.Namespace))
+	if err != nil {
+		return nil, err
+	}
+	var connectedClients []connectedClient
+	if err = yaml.Unmarshal([]byte(clients), connectedClients); err != nil {
+		return nil, err
+	}
```

Comment:
I don't get what you are doing here. 
Why are we even looking at information that does not come directly from this storage consumers list?

```suggestion
connectedClients := make([]connectedClient, 0, len(storageConsumers))
```

The `len` as the third param will gureentee that the appends in the loop below will not relocate and will not reallocate any memory

Reply:
See [here](https://github.com/raaizik/ocs-operator/blob/ad46b03c29e9c82c1496254aaa1537a65783e012/controllers/storagecluster/odfinfoconfig.go#L85): I'm fetching the clients as a string and passing it down to this function for the case of an already existing CM that needs to have its clients array appended.

Reply:
You don't append anything to an existing list. You regenerate the list from the storage consumer list. The storage consumer list is the source of truth and on each reconcile we are re-building the CM data including the client list from the ground up. Of course if the resulted CM is identical to the old CM, CreateOrUpdate will NOT issue any k8s API calls